Pass Jaeger references with a 0 SpanID and TraceID

joe-elliott opened the pull request:
**Description:**
This PR includes targeted changes to make the Jaeger pipeline tolerate references that have a Trace/Span ID of 0.

When translating from ocproto to jaeger a nil TraceID is returned and stored due to this code:
https://github.com/open-telemetry/opentelemetry-collector/blob/34b2459597f12409d83c20ab3ec8414e28221d81/translator/trace/big_endian_converter.go#L35-L38

When the reverse translation was performed the translation failed with `ErrNilTraceID`:
https://github.com/open-telemetry/opentelemetry-collector/blob/34b2459597f12409d83c20ab3ec8414e28221d81/translator/trace/big_endian_converter.go#L53-L56

When the `ErrNilTraceID` and `ErrNilSpanID` are returned 0's are used instead of throwing an error. 
 Changes have been tested manually to successfully pass bad references in a way consistent with the Jaeger pipeline/backend.

Fixes #491 

**Testing:**
A reference with a 0 span and trace id was added to the protospan -> jaegerproto and protospan -> jaegerthrift tests.

owais commented on translator/trace/jaeger/protospan_to_jaegerproto.go:
```diff
@@ -187,7 +187,10 @@ func ocLinksToJaegerReferencesProto(ocSpanLinks *tracepb.Span_Links) ([]jaeger.S
-		if err != nil {
+		if err == tracetranslator.ErrNilTraceID {
+			traceIDLow = 0
+			traceIDHigh = 0
```
Does this only apply to jaeger? If we are fine with this behavior, may be this function should not return an error at all and accept 0 as a valid ID?

joe-elliott replied:
I had considered this, but `BytesToInt64TraceID` and `BytesToUInt64TraceID` are also used to translate the TraceID on the span itself and I did not want to adjust this behavior.

I'm honestly not sure what the Jaeger pipeline would do with a span with a 0 trace or span ID.

tigrannajaryan replied:
I think we should adopt this behavior universally for all translation code (not just Jaeger).

If we receive a span with 0 ID we should pass it as is (if we can). It is up to Jaeger to decide how they deal with a 0 ID span that they receive from the Collector. Collector is not intended to be a sanitizer of telemetry data _by default_. The default behavior should be to pass everything as is, without modifying anything unless the modification cannot be avoided. This should be the default behavior and a design goal for the Collector. It is not Collector's business to interpret and modify data _unless_ the end user specifically configures it to do so using processors.

If sanitizing of 0 ID is desirable then it can be implemented as a processor.

owais replied:
Agree with Tigran. That's what I was leading to as well.

joe-elliott replied:
Thanks for the direction.  PR has been updated.  See the latest comment.

joe-elliott commented:
@tigrannajaryan @owais 
Taking our conversation into account the PR has been updated.  The ErrNilSpan/TraceID errors have been removed and instead 0's are returned.  This is a low impact change that, I believe, accomplishes our goals.

There is another approach that I like a bit more conceptually, but will result in more code changes:  
- Leave `ErrNilTraceID` and `ErrNilSpanID`.
- Change `UInt64ToByteTraceID` and `UInt64ToByteSpanID` to return valid TraceID's when 0's are passed in instead of nil.  i.e. return a `[]byte` full of 0x00's.
- Adjust impacted translation logic accordingly.

Thoughts?

tigrannajaryan commented:
> @tigrannajaryan @owais
> Taking our conversation into account the PR has been updated. The ErrNilSpan/TraceID errors have been removed and instead 0's are returned. This is a low impact change that, I believe, accomplishes our goals.
> 
> There is another approach that I like a bit more conceptually, but will result in more code changes:
> 
>     * Leave `ErrNilTraceID` and `ErrNilSpanID`.
> 
>     * Change `UInt64ToByteTraceID` and `UInt64ToByteSpanID` to return valid TraceID's when 0's are passed in instead of nil.  i.e. return a `[]byte` full of 0x00's.
> 
>     * Adjust impacted translation logic accordingly.
> 
> 
> Thoughts?

This is what I would do. Unclear why the code is not already written this way. I don't see why 0 trace ID is treated specially. 0 uint64 should convert to byte sequence of zeros and vice versa and that would happen if there were no explicit `if`s that check for 0 uint64.

Perhaps there are edge cases where this matters? @owais are you aware of any?

owais commented:
No, don't know of any issues this could cause.

tigrannajaryan commented:
@pjanotti any thoughts on this? https://github.com/open-telemetry/opentelemetry-collector/pull/513#issuecomment-577462151

joe-elliott commented:
@tigrannajaryan @owais 
Third time's the charm!

- Removed logic that returns nil when trace/span id of 0 is requested
- Added tests to confirm 0 trace/span id maps to slices of 0x00 bytes
- Adjusted the way ParentSpanID's were translated.  Previously the code had relied on the return of a nil slice to have an empty ParentSpanID
- Adjusted tests that relied on the nil slice instead of 0's

tigrannajaryan commented on translator/trace/jaeger/jaegerproto_to_protospan.go:
```diff
@@ -101,18 +101,21 @@ func jProtoSpansToOCProtoSpans(jspans []*model.Span) []*tracepb.Span {
 		}
 
+		if jspan.ParentSpanID() != 0 {
+			span.ParentSpanId = tracetranslator.UInt64ToByteSpanID(uint64(jspan.ParentSpanID()))
```
Is this special-cased because we want ParentSpanId in OC format to be nil when ParentSpanID in Jaeger format is 0?

tigrannajaryan replied:
From what I see we already have a special-case in the reverse translation [1] so this looks good to me.

[1] https://github.com/open-telemetry/opentelemetry-collector/blob/9ecbe75898683aa5316a80ae1a5cb15dfdd1651f/translator/trace/jaeger/protospan_to_jaegerproto.go#L492

joe-elliott replied:
So whenever we translate from a 0 trace/span ID we have to decide if that means a 0 id or a nil id.  We've decided we feel like the default is all 0's which I like, but ParentSpanId is a special case because a nil ParentSpanId is common and correct.

There is a special case on the reverse translation back to Jaeger, but if we don't translate back to Jaeger then I believe we would want clearly flag "no parent id" with a nil instead of a 0's byte slice which might be confusing.

Another option is to go with back to my second attempt which was basically:
- Translate all 0 span/trace ids as nil
- When you translate any nil span/trace id back to Jaeger choose 0.

tigrannajaryan replied:
I think this PR is good, no need to go back to the previous attempt.

pjanotti commented:
> @pjanotti any thoughts on this? #513 (comment)

@tigrannajaryan I don't recall any good reason for those...

pjanotti commented:
@tigrannajaryan @joe-elliott ah I think it was because of this https://github.com/census-instrumentation/opencensus-proto/blob/master/gen-go/trace/v1/trace.pb.go#L139. That said this should not be purged in the pipeline by default (I already had some issues with it in the past). If someone wishes to enforce this OC policy they should not rely on the pipeline to do that. It can be a processor or an option on receiver/exporter. So I'm merging it :)
